### DISCUSSION TITLE
Add support for Home Storage Batteries & Evs

### DISCUSSION THREAD

**Discussion by swinbsp** [andrei-tatar/node-red-contrib-smartnora#71] [Ideas]
Storging PV energy, and grid leveling via time of use tariffs is a growing market. Very very few batteries have any smart assistant based controls, however a number do have APIs to monitor and control. It would be great to expose these in Google Home and automate via NodeRed  ( I already do the automations for my system )

Google have these classifies as "chargers"

https://developers.google.com/assistant/smarthome/guides/charger

I beleive from the post relating to security devices you have already made progress on gerneral traits and commands. This would be a specific extension of that generic capability

Key monitoring data is SOC ( status of Chrage ), Charging Status ( charging, idle, discharging)

Key commands are Charge and Discharge the battery

Whilst not explicitly in catagory, the same control can be used for EV (electric cars). Again they can easily be controlled via APIs but are not gernerally made avaialble in smart assistants

**Comment by andrei-tatar** [comment]
Sounds like a useful thing

**Comment by arb2040** [comment]
I was hoping to find a node that (at least) can display the state-of-charge of my EV, but cannot find a good alternative.  I can use Humidity to display percent, but it falls under Climate devices - so not quite right. 
I wish for a node (I know it may be limited by Google) where I can see if the car is connected and/or charging - in addition to the SoC.  Any hope for this?

**Reply by andrei-tatar** [reply]
@arb2040 There is the `charger` node (which doesn't have any documentation yet) which supports the EnergyStorage trait https://developers.home.google.com/cloud-to-cloud/traits/energystorage .
You can expose the SoC in percentage and a boolean that indicates if it's charging or not (`isCharging`) or if the EV is connected (`isPluggedIn`).
Not sure how it appears in Google Home, but it should work with voice queries.

**Reply by andrei-tatar** [reply]
@arb2040 I added the node to my flow to report on my home battery status. This is the flow I'm using to report the state of charge in %

```json
[{"id":"860d818f14cd1779","type":"noraf-charger","z":"d02cf25fa8114475","devicename":"Home Battery","roomhint":"Service Room","name":"","passthru":false,"nora":"f1e538c1.0315f8","isRechargeable":true,"queryOnlyEnergyStorage":true,"energyStorageDistanceUnitForUX":"KILOMETERS","topic":"","twofactor":"off","twofactorpin":"","filter":false,"asyncCmd":false,"outputs":1,"x":940,"y":100,"wires":[[]]},{"id":"8bb4278aca29cc84","type":"function","z":"d02cf25fa8114475","name":"map","func":"const batteryCapacity = global.get('batteryCapacity');\n\nconst isCharging = flow.get('batteryStatus') === 'charge';\n\nconst capacityRemaining = [{\n    rawValue: batteryCapacity,\n    unit: 'PERCENTAGE',\n}];\n\nconst capacityUntilFull = [{\n    rawValue: 100 - batteryCapacity,\n    unit: 'PERCENTAGE',\n}];\n\nconst descriptiveCapacityRemaining = [\n    { threshold: 98, message: 'FULL' },\n    { threshold: 80, message: 'HIGH' },\n    { threshold: 40, message: 'MEDIUM' },\n    { threshold: 20, message: 'LOW' },\n    { threshold: 0, message: 'CRITICALLY_LOW' },\n]\n    .find(v => batteryCapacity >= v.threshold)\n    .message;\n\nreturn {\n    payload: {\n        capacityRemaining,\n        capacityUntilFull,\n        descriptiveCapacityRemaining,\n        isCharging,\n        isPluggedIn: true,\n    },\n};","outputs":1,"timeout":0,"noerr":0,"initialize":"","finalize":"","libs":[],"x":815,"y":100,"wires":[["860d818f14cd1779"]],"l":false},{"id":"f1e538c1.0315f8","type":"noraf-config","name":"NORA","group":"seceani","twofactor":"off","twofactorpin":"","localexecution":true,"structure":"Seceani","storeStateInContext":true,"disableValidationErrors":false,"sendDeviceNameAndLocation":false}]
```
